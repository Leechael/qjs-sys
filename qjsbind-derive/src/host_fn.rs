use proc_macro2::TokenStream;
use template_quote::quote;

pub(crate) fn patch(input: TokenStream, with_context: bool) -> TokenStream {
    match patch_or_err(input, with_context) {
        Ok(tokens) => tokens,
        Err(err) => err.to_compile_error(),
    }
}

fn patch_or_err(input: TokenStream, with_context: bool) -> syn::Result<TokenStream> {
    let the_fn: syn::ItemFn = syn::parse2(input.clone())?;
    let fn_ident = &the_fn.sig.ident;
    let crate_qjsbind = crate::find_crate_name("qjsbind")?;
    let args = the_fn.sig.inputs;
    let arg_names = args
        .iter()
        .filter_map(|arg| match arg {
            syn::FnArg::Typed(pat) => match &*pat.pat {
                syn::Pat::Ident(ref ident) => Some(ident.ident.clone()),
                _ => None,
            },
            _ => None,
        })
        .collect::<Vec<_>>();
    let output = match the_fn.sig.output {
        syn::ReturnType::Default => quote! { () },
        syn::ReturnType::Type(_, ty) => quote! { #ty },
    };
    Ok(quote! {
        pub unsafe extern "C" fn #fn_ident(
            ctx: *mut #crate_qjsbind::c::JSContext,
            this_val: #crate_qjsbind::c::JSValueConst,
            argc: core::ffi::c_int,
            argv: *mut #crate_qjsbind::c::JSValue,
        ) -> #crate_qjsbind::c::JSValue
        {
            #input
            #(if with_context) {
                #crate_qjsbind::call_host_function(#fn_ident, ctx, this_val, argc, argv)
            }
            #(else) {
                fn wrapper(_ctx: #crate_qjsbind::Context, _this: #crate_qjsbind::Value, #args) -> #output {
                    #fn_ident(#(#arg_names),*)
                }
                #crate_qjsbind::call_host_function(wrapper, ctx, this_val, argc, argv)
            }
        }
    })
}
