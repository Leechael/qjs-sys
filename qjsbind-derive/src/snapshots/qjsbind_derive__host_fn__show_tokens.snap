---
source: qjsbind-derive/src/host_fn.rs
assertion_line: 115
expression: "rustfmt_snippet::rustfmt(&patched.to_string()).unwrap()"
---
pub unsafe extern "C" fn qjsbind_CryptoKey_constructor(
    c_ctx: *mut qjsbind::c::JSContext,
    c_this: qjsbind::c::JSValueConst,
    argc: core::ffi::c_int,
    argv: *mut qjsbind::c::JSValue,
) -> qjsbind::c::JSValue {
    fn qjsbind_CryptoKey_constructor(
        ctx: crate_js::Context,
        _this_value: crate_js::Value,
        inner: CryptoKey,
    ) -> crate_js::Result<crate_js::Native<CryptoKey>> {
        #[allow(unused_variables)]
        let ctx = ctx;
        use crate_js::IntoNativeObject;
        CryptoKey::new(inner).into_native_object(&ctx)
    }
    #[allow(unused_variables)]
    let ctx =
        qjsbind::Context::clone_from_ptr(c_ctx).expect("calling host function with null context");
    let args = unsafe { core::slice::from_raw_parts(argv, argc as usize) };
    let mut args = args
        .into_iter()
        .map(|v| qjsbind::Value::new_cloned(&ctx, *v));
    let this_value = qjsbind::Value::new_cloned(&ctx, c_this);
    let rv: qjsbind::Result<_> = {
        let ctx = ctx.clone();
        (move || {
            Ok(qjsbind_CryptoKey_constructor(
                ctx.try_into()?,
                qjsbind::FromJsValue::from_js_value(this_value)?,
                qjsbind::FromJsValue::from_js_value(
                    args.next().unwrap_or(qjsbind::Value::undefined()),
                )?,
            ))
        })()
    };
    qjsbind::convert_host_call_result("qjsbind_CryptoKey_constructor", &ctx, rv)
}
